Update loader documentation. (#685)
KhronosGroup/Vulkan-Loader: docs/LoaderInterfaceArchitecture.md

corresponding type.
For example, an "instance extension" is composed primarily of "instance
functions" which primarily take instance objects.
These will be discussed in more detail later.

#### Device-Specific

A Vulkan device (`VkDevice`), on the other-hand, is a logical identifier used
to associate functions with a particular Vulkan physical device
(`VkPhysicalDevice`) through a particular implementation on a user's system.

##### Device Objects

A few of the Vulkan constructs associated directly with a device include:
 * `VkDevice`
 * `VkQueue`
 * `VkCommandBuffer`
 * Any dispatchable object that is a child of one of the above

##### Device Functions

A "device function" is any Vulkan function which takes any device object as its
first parameter.
The vast majority of Vulkan functions are device functions.
Some Vulkan device functions are:
 * `vkQueueSubmit`
 * `vkBeginCommandBuffer`
 * `vkCreateEvent`

Vulkan devices functions may be queried using either `vkGetInstanceProcAddr` or
`vkGetDeviceProcAddr`.
If an application chooses to use `vkGetInstanceProcAddr`, each call will have
additional function calls built into the call chain, which will reduce performance
slightly.

> However, the function pointer returned can be used for any device created later,

Author: I removed that sentence (the one that starts with "However").  I don't think it added anything.

   - Created in the loader during the call to `vkCreateInstance`
 - Device Dispatch Table
   - Created in the loader during the call to `vkCreateDevice`

At that time the application and the system can each specify optional layers to
be included.
The loader will initialize the specified layers to create a call chain for each
Vulkan function and each entry of the dispatch table will point to the first
element of that chain.
Thus, the loader builds an instance call chain for each `VkInstance` that is
created and a device call chain for each `VkDevice` that is created.

When an application calls a Vulkan function, this typically will first hit a
*trampoline* function in the loader.
These *trampoline* functions are small, simple functions that jump to the
appropriate dispatch table entry for the object they are given.
Additionally, for functions in the instance call chain, the loader has an
additional function, called a *terminator*, which is called after all enabled
layers to marshall the appropriate information to all available implementations.

#### Instance Call Chain Example

For example, the diagram below represents what happens in the call chain for
`vkCreateInstance`.

> After initializing the chain, the loader will call into the first layer's
> `vkCreateInstance` which will call the next finally terminating in the loader
> again where this function calls every implementation's `vkCreateInstance` and
> saves the results.

Reviewer: ```suggestion
After initializing the chain, the loader calls into the first layer's
`vkCreateInstance`, which will call the next layer's `vkCreateInstance 
before finally terminating in the loader again where it will call
every implementation's `vkCreateInstance`.
```

> This allows every enabled layer for this chain to set up what it needs based on
> the `VkInstanceCreateInfo` structure from the application.

Reviewer: ```suggestion
This allows every enabled layer in the chain to set up what it needs based on
the `VkInstanceCreateInfo` structure from the application.
```

![Instance Call Chain](./images/loader_instance_chain.png)

This also highlights some of the complexity the loader must manage when using
instance call chains.
As shown here, the loader's *terminator* must aggregate information to and from
multiple implementations when they are present.
This implies that the loader has to be aware of any instance-level extensions
which work on a `VkInstance` to aggregate them correctly.

#### Device Call Chain Example

> Device call chains are created at `vkCreateDevice` and are generally simpler
> because they deal with only a single device and the implementation can always be
> the *terminator* of the chain.

Reviewer: ```suggestion
Device call chains are created in `vkCreateDevice` and are generally simpler
because they deal with only a single device. This allows for the implementation 
to always be the *terminator* of the chain.
```

![Loader Device Call Chain](./images/loader_device_chain_loader.png)
<br/>

## Elevated Privilege Caveats

> Normally, Vulkan applications should not be run using elevated privileges,
> this is to keep the system safe from exploitation.
> However, should the need arise, some behaviors of the loader are different when
> running a Vulkan application in an elevated fashion.
> This is done to ensure that an application with elevated privileges does
> not run using components that were not installed in proper approved locations.

Reviewer: ```suggestion
To ensure that the system is safe from exploitation, Vulkan applications which
are run with elevated privileges are restricted from certain operations, such
as reading environment variables from unsecure locations or searching for
files in user controlled paths. This is done to ensure that an application
running with elevated privileges does not run using components that were
not installed in the proper approved locations. 
```

> Additionally, the loader uses `secure_getenv` (or the platform-specific
> equivalent) to query sensitive environment variables to avoid accidentally using
> untrusted results.

Reviewer: ```suggestion
The loader uses platform-specific mechanisms (such as `secure_getenv` and its
equivalents) for querying sensitive environment variables to avoid accidentally using
untrusted results.
```

<br/>
<br/>

## Application Interface to the Loader

The Application interface to the Vulkan loader is now detailed in the
[LoaderApplicationInterface.md](LoaderApplicationInterface.md) document found in
the same directory as this file.
<br/>
<br/>

## Layer Interface with the Loader

The Layer interface to the Vulkan loader is detailed in the
[LoaderLayerInterface.md](LoaderLayerInterface.md) document found in the same
directory as this file.
<br/>
<br/>

## Implementation Interface With the Loader

The Implementation interface to the Vulkan loader is detailed in the
[LoaderImplementationInterface.md](LoaderImplementationInterface.md) document
found in the same directory as this file.
<br/>
<br/>

## Table of Debug Environment Variables

The following are all the Debug Environment Variables available for use with the
Loader.
These are referenced throughout the text, but collected here for ease of
discovery.

<table style="width:100%">
  <tr>
    <th>Environment Variable</th>
    <th>Behavior</th>
    <th>Example Format</th>
  </tr>
  <tr>
    <td><small><i>VK_ICD_FILENAMES</i></small></td>
    <td>Force the loader to use the specific ICD JSON files.

>         The value should contain a list of delimited full path listings to
>         implementation JSON Manifest files.<br/>

Reviewer: ```suggestion
        The value contains a list of delimited full path listings to
        implementation JSON Manifest files.<br/>
```

        &nbsp;&nbsp;VK_LAYER_PATH=<br/>
        &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&lt;path_a&gt;;&lt;path_b&gt;<br/><br/>
        set<br/>
        &nbsp;&nbsp;VK_LAYER_PATH=<br/>
        &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&lt;path_a&gt;;&lt;path_b&gt;</small>
    </td>
  </tr>
  <tr>
    <td><small><i>VK_LOADER_DISABLE_INST_EXT_FILTER</i></small></td>
    <td>Disable the filtering out of instance extensions that the loader doesn't
        know about.
        This will allow applications to enable instance extensions exposed by
        implementations but that the loader has no support for.<br/>
        <b>NOTE:</b> This may cause the loader or application to crash.</td>
    <td><small>export<br/>
        &nbsp;&nbsp;VK_LOADER_DISABLE_INST_EXT_FILTER=1<br/><br/>
        set<br/>
        &nbsp;&nbsp;VK_LOADER_DISABLE_INST_EXT_FILTER=1</small>
    </td>
  </tr>
  <tr>
    <td><small><i>VK_LOADER_DEBUG</i></small></td>
    <td>Enable loader debug messages using a comma-delimited list of level
        options.  These options are:<br/>
        &nbsp;&nbsp;* error (only errors)<br/>
        &nbsp;&nbsp;* warn (only warnings)<br/>
        &nbsp;&nbsp;* info (only info)<br/>

>         &nbsp;&nbsp;* debug (only debug)<br/>

Reviewer: No love for `layer` level info?

        error messages, set the value to "error,warn,info".
    </td>
    <td><small>export<br/>
        &nbsp;&nbsp;VK_LOADER_DEBUG=all<br/>
        <br/>
        set<br/>
        &nbsp;&nbsp;VK_LOADER_DEBUG=warn</small>
    </td>
  </tr>
</table>
<br/>
<br/>

## Glossary of Terms

<table style="width:100%">
  <tr>
    <th>Field Name</th>
    <th>Field Value</th>
  </tr>
  <tr>
    <td>Android Loader</td>
    <td>The loader designed to work primarily for the Android OS.
        This is generated from a different code base than the desktop loader.
        But, in all important aspects, it should be functionally equivalent.
    </td>
  </tr>
  <tr>
    <td>Desktop Loader</td>
    <td>The loader designed to work on Windows, Linux and macOS.
        This is generated from a different
        <a href="https://github.com/KhronosGroup/Vulkan-Loader">code base</a>
        than the Android loader.
        But in all important aspects, it should be functionally equivalent.
    </td>
  </tr>
  <tr>
    <td>Core Function</td>
    <td>A function that is already part of the Vulkan core specification and not
        an extension. <br/>

>         For example, <b>vkCreateDevice()</b>.

Reviewer: Is it worth adding the description "Extension functions always have a suffix to denote which type of extension they are, such as KHR and EXT"?

        available. <br/>
        Always first query if an extension exists, and enable it during
        <b>vkCreateInstance</b> (if it is an instance extension) or during
        <b>vkCreateDevice</b> (if it is a device extension) before attempting
        to use it.
    </td>
  </tr>
  <tr>
    <td>ICD</td>
    <td>Acronym for "Installable Client Driver".
        These are drivers that are provided by IHVs to interact with the
        hardware they provide. <br/>
        These are the most common type of Vulkan implementations. <br/>
        See <a href="#installable-client-drivers">Installable Client Drivers</a> 
        section for more information.
    </td>
  </tr>
  <tr>
    <td>IHV</td>
    <td>Acronym for an "Independent Hardware Vendor".
        Typically the company that built the underlying hardware technology
        that is being used. <br/>
        A typical examples for a Graphics IHV include (but not limited to):
        AMD, ARM, Imagination, Intel, Nvidia, Qualcomm
    </td>
  </tr>
  <tr>
    <td>Implementation</td>
    <td>The underlying library which provides support for the Vulkan API.

>         This support can be implemented as either an ICD, graphics conversion
>         library, or pure software.<br/>

Reviewer: ```suggestion
        This support can be implemented as either an ICD, API translation
        library, or pure software.<br/>
```

        trampoline calls enabled layers, the final layer calls a loader
        terminator, and the loader terminator calls all available
        implementations. <br/>
        See the <a href="#dispatch-tables-and-call-chains">Dispatch Tables and
        Call Chains</a> section for more information.
    </td>
  </tr>
  <tr>
    <td>Instance Function</td>
    <td>An instance function is any Vulkan function which takes as its first
        parameter either a <i>VkInstance</i> or a <i>VkPhysicalDevice</i> or
        nothing at all. <br/><br/>
        Some Vulkan instance functions are:<br/>
        &nbsp;&nbsp;<b>vkEnumerateInstanceExtensionProperties</b>, <br/>
        &nbsp;&nbsp;<b>vkEnumeratePhysicalDevices</b>, <br/>
        &nbsp;&nbsp;<b>vkCreateInstance</b>, <br/>
        &nbsp;&nbsp;<b>vkDestroyInstance</b>. <br/><br/>
        See the <a href="#instance-versus-device">Instance Versus Device</a>
        section for more information.
    </td>
  </tr>
  <tr>
    <td>Layer</td>
    <td>Layers are optional components that augment the Vulkan system.
        They can intercept, evaluate, and modify existing Vulkan functions on

>         their way from the application down to the hardware.<br/>

Reviewer: ```suggestion
        their way from the application down to the implementation.<br/>
```

        See the <a href="#layers">Layers</a> section for more information.
    </td>
  </tr>
  <tr>
    <td>Loader</td>
    <td>The middleware program which acts as the mediator between Vulkan

>         applications, Vulkan layers and Vulkan drivers.<br/>

Reviewer: ```suggestion
        applications, Vulkan layers and, Vulkan implementations.<br/>
```

        See <a href="#the-loader">The Loader</a> section for more information.
    </td>
  </tr>
  <tr>
    <td>Manifest Files</td>
    <td>Data files in JSON format used by the desktop loader.
        These files contain specific information for either a
        <a href="LoaderLayerInterface.md#layer-manifest-file-format">Layer</a>
        or an

>         <a href="LoaderImplementationInterface.md#icd-manifest-file-format">Implementation</a> and
>         define where to find necessary files and default settings.

Reviewer: ```suggestion
        <a href="LoaderImplementationInterface.md#icd-manifest-file-format">Implementation</a> and
        define necessary information such as where to find files and default settings.
```
